File: 2024/q13.py
#!/usr/bin/env python
import heapq
import io


def display(grid, path):
    coff = '\033[0m'
    con = '\033[31m'
    minx = min(x for x,y in grid)
    maxx = max(x for x,y in grid)
    miny = min(y for x,y in grid)
    maxy = max(y for x,y in grid)
    with io.StringIO() as so:
        for y in range(miny, maxy+1):
            for x in range(minx, maxx+1):
                if (c := grid.get((x,y))) is not None:
                    if (x,y) in path:
                        print(con, end='', file=so)
                        print(c, end='', file=so)
                        print(coff, end='', file=so)
                    else:
                        print(c, end='', file=so)
                else:
                    print(' ', end='', file=so)
            print(file=so)
        print(so.getvalue())


def part1(data):
    abc = 'S0123456789E'
    grid = {(x,y):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s) if c in abc}
    S = next(p for p,c in grid.items() if c == 'S')
    E = next(p for p,c in grid.items() if c == 'E')
    grid = {p:(0 if c in'SE' else int(c)) for p,c in grid.items()}
    ms = [(1,0), (-1,0), (0,1), (0,-1)]
    h0 = [0, 1, 2, 3, 4, 5, 4, 3, 2, 1]
    ans = 0
    fringe = [(0, S, (S,))]
    seen = set()
    while fringe:
        wei, pos, path = heapq.heappop(fringe)
        if pos == E:
            ans = wei
            break
        if pos in seen: continue
        seen.add(pos)
        h = grid[pos]
        x,y = pos
        for dx,dy in ms:
            q = (x+dx,y+dy)
            if (l := grid.get(q)) is not None:
                w = h0[(h-l) % len(h0)]
                heapq.heappush(fringe, (wei + w + 1, q, path + (q,)))
    return ans


def part2(data):
    return part1(data)


def part3(data):
    abc = 'S0123456789E'
    grid = {(x,y):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s) if c in abc}
    S = [p for p,c in grid.items() if c == 'S']
    E = next(p for p,c in grid.items() if c == 'E')
    grid = {p:(0 if c in'SE' else int(c)) for p,c in grid.items()}
    ms = [(1,0), (-1,0), (0,1), (0,-1)]
    h0 = [0, 1, 2, 3, 4, 5, 4, 3, 2, 1]
    ans = 0
    fringe = [(0, s, (s,)) for s in S]
    heapq.heapify(fringe)
    seen = set()
    while fringe:
        wei, pos, path = heapq.heappop(fringe)
        if pos == E:
            ans = wei
            break
        if pos in seen: continue
        seen.add(pos)
        h = grid[pos]
        x,y = pos
        for dx,dy in ms:
            q = (x+dx,y+dy)
            if (l := grid.get(q)) is not None:
                w = h0[(h-l) % len(h0)]
                heapq.heappush(fringe, (wei + w + 1, q, path + (q,)))
    return ans


data = """
#######
#6769##
S50505E
#97434#
#######
"""
assert part1(data) == 28

data = """
SSSSSSSSSSS
S674345621S
S###6#4#18S
S53#6#4532S
S5450E0485S
S##7154532S
S2##314#18S
S971595#34S
SSSSSSSSSSS
"""
assert part3(data) == 14


data = open('q13_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q13_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q13_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: event/2024/q13.jl
#!/usr/bin/env julia
using DataStructures


const Pos = CartesianIndex{2}


function parsegrid(text)
    s = stack(split(text, '\n', keepempty=false), dims=1)
    start = findall(==('S'), s)
    goal = findfirst(==('E'), s)
    d = Dict(zip(['S';'E';'0':'9'], [0;0;0:9]))
    grid = [get(d, c, missing) for c in s]
    return (grid, start, goal)
end


function printgrid(grid, pois)
    con, coff = "\033[31m", "\033[0m"
    println()
    for y in axes(grid, 1)
        s = ""
        for x in axes(grid, 2)
            p = Pos(y, x)
            c = ismissing(grid[p]) ? ' ' : ('0'+grid[p])
            s *= p in pois ? string(con,c,coff) : c
        end
        println(s)
    end
end


function unwind(i, paths)
    seen = Set{Pos}()
    while i != 0
        i, pos = paths[i]
        push!(seen, pos)
    end
    return seen
end


function walk(grid, start, goal)
    neibs = [(0,1),(-1,0),(1,0),(0,-1)] .|> Pos
    h = [1,2,3,4,5,4,3,2,1,0]
    fringe = BinaryMinHeap([(0, grid[start], start)])
    seen = Set{Pos}()
    while !isempty(fringe)
        dist, level, pos = pop!(fringe)
        pos == goal && return dist
        pos in seen && continue
        push!(seen, pos)
        for q in pos .+ neibs
            c = get(grid, q, missing)
            ismissing(c) && continue
            w = 1 + h[mod(c - level, 1:10)]
            push!(fringe, (dist+w, c, q))
        end
    end
end


function part1(data)
    grid, (start,), goal = parsegrid(data)
    walk(grid, start, goal)
end


part2 = part1


function part3(data)
    grid, start, goal = parsegrid(data)
    walk.([grid], start, goal) |> minimum
end


data = raw"
#######
#6769##
S50505E
#97434#
#######
"
@assert part1(data) == 28


data = raw"
SSSSSSSSSSS
S674345621S
S###6#4#18S
S53#6#4532S
S5450E0485S
S##7154532S
S2##314#18S
S971595#34S
SSSSSSSSSSS
"
@assert part3(data) == 14


data = readchomp("q13_p1.txt")
println("part1: ", part1(data))
data = readchomp("q13_p2.txt")
println("part2: ", part2(data))
data = readchomp("q13_p3.txt")
println("part3: ", part3(data))
